File: OsmAnd/src/net/osmand/plus/helpers/Kml2Gpx.java
package net.osmand.plus.helpers;

import android.annotation.TargetApi;
import net.osmand.PlatformUtil;
import org.apache.commons.logging.Log;

import javax.xml.transform.*;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;
import java.io.InputStream;
import java.io.StringWriter;

/**
 * @author Koen Rabaey
 */
public class Kml2Gpx {

	public static final Log LOG = PlatformUtil.getLog(Kml2Gpx.class);

	@TargetApi(8)
	public static String toGpx(InputStream kml) {
		try {
			Source xmlSource = new StreamSource(kml);
			Source xsltSource = new StreamSource(Kml2Gpx.class.getResourceAsStream("kml2gpx.xslt"));

			StringWriter sw = new StringWriter();

			TransformerFactory.newInstance().newTransformer(xsltSource).transform(xmlSource, new StreamResult(sw));

			return sw.toString();

		} catch (TransformerConfigurationException e) {
			LOG.error(e.toString(), e);
		} catch (TransformerFactoryConfigurationError e) {
			LOG.error(e.toString(), e);
		} catch (TransformerException e) {
			LOG.error(e.toString(), e);
		}

		return null;
	}
}

File: OsmAnd-shared/src/jvmMain/kotlin/net/osmand/shared/gpx/helper/Kml2Gpx.kt
package net.osmand.shared.gpx.helper

import net.osmand.shared.io.SourceInputStream
import net.osmand.shared.util.LoggerFactory
import java.io.StringWriter
import javax.xml.transform.Source
import javax.xml.transform.TransformerException
import javax.xml.transform.TransformerFactory
import javax.xml.transform.TransformerFactoryConfigurationError
import javax.xml.transform.stream.StreamResult
import javax.xml.transform.stream.StreamSource

actual object Kml2Gpx {
	private val LOG = LoggerFactory.getLogger(Kml2Gpx::class.simpleName!!)

	actual fun toGpx(kml: okio.Source): String? {
		try {
			val xmlSource: Source = StreamSource(SourceInputStream(kml))
			val xsltSource: Source = StreamSource(this::class.java.getResourceAsStream("/kml2gpx.xslt"))

			val sw = StringWriter()

			TransformerFactory.newInstance().newTransformer(xsltSource).transform(xmlSource, StreamResult(sw))

			return sw.toString()
		} catch (e: TransformerFactoryConfigurationError) {
			LOG.error(e.toString(), e)
		} catch (e: TransformerException) {
			LOG.error(e.toString(), e)
		}
		return null
	}
}
